streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java

```diff
@@ -116,6 +116,7 @@ private void setup(final boolean enableCache) {
                 results.add(new KeyValueTimestamp<>(record.key(), record.value(), record.timestamp()));
             }
         };
+        context.setTime(0L);
```

Comment:
Interesting... Why did this need to be changed?

Reply:
Good catch. This test uses `CachingSessionStore`. 

It fails when caching is enabled because `SessionStore` is caching based on a context passed on initialization, not record's one:

```java
    private void initInternal(final InternalProcessorContext<?, ?> context) {
        this.context = context;
        //...
    }

//...
    @Override
    public void put(final Windowed<Bytes> key, final byte[] value) {
        validateStoreOpen();
        final Bytes binaryKey = SessionKeySchema.toBinary(key);
        final LRUCacheEntry entry =
            new LRUCacheEntry(
                value,
                context.headers(),
                true,
                context.offset(),
                context.timestamp(),
                context.partition(),
                context.topic());
        context.cache().put(cacheName, cacheFunction.cacheKey(binaryKey), entry);

        maxObservedTimestamp = Math.max(keySchema.segmentTimestamp(binaryKey), maxObservedTimestamp);
    }
```

When flushing and context is load from cache, it gets the `timestamp=-1`:

```java
                final ProcessorRecordContext current = context.recordContext();
                context.setRecordContext(entry.entry().context());
                try {
                    flushListener.apply(
                        new Record<>(
                            binaryKey.get(),
                            new Change<>(newValueBytes, sendOldValues ? oldValueBytes : null),
                            entry.entry().context().timestamp(),
                            entry.entry().context().headers()));
                } finally {
                    context.setRecordContext(current);
                }

```

Before this timestamp was passed from the key, but now that we are building a record, its validation is failing.

This could be valid workaround but we could also look into `StateStore`s to use records instead of K/V. wdyt?

Reply:
Ah, ok, I finally got it. What was weirding me out is that the context's timestamp should actually always match the record's timestamp. When we create a Record and pass it in to the first processor in the topology, we create the Record using the context's timestamp, and then when we forward a record in ProcessorContextImpl, we override the context's timestamp based on the forwarded record's timestamp. So, it shouldn't be possible to have a valid record timestamp go in to the cache, but an invalid timestamp come out, regardless of whether the cache is using the context or the record.

However, this is a unit test, and we're directly invoking the processor, bypassing the logic in the ProcessorContextImpl to make the context and record agree. So all bets are off. Even before this PR, the cache was returning an invalid -1 timestamp, but no one ever checked it. The difference is that now, we're constructing a Record, which checks that its timestamp is valid.

I think your patchup is fine, again, because the result timestamp doesn't matter, and the cache isn't the unit under test in this unit test. If you wanted to do a "proper" fix, you could insert a method like `invokeProcessor(Record)`, which would first set the timestamp on the context and then invoke the Processor.

By the way, I just wrote and erased a whole section because I was afraid this change might actually hurt users whose unit tests might just start to mysteriously fail like this. I was going to suggest that we throw an exception in the InternalMockProcessorContext#timestamp if the timestamp hasn't been initialized:
```
    private long timestamp = -1L;

...

    @Override
    public long timestamp() {
        if (recordContext == null) {
            return timestamp;
        }
        return recordContext.timestamp();
    }
```

Then I realized that that's the internal mock... Guess what, the public mock context already has that check:
```
    @Override
    public long timestamp() {
        if (recordTimestamp == null) {
            throw new IllegalStateException("Timestamp must be set before use via setRecordMetadata() or setTimestamp().");
        }
        return recordTimestamp;
    }
```

Too bad we're not as nice to ourselves as we are to our users!

Anyway, I'm fine with this small fix. If you also wanted to remove this little bear trap for the next contributor, you might consider adding that sanity check to the internal mock context.

Reply:
Agree.

Here's another interesting finding: 

https://github.com/apache/kafka/blob/4fd71a7ef1c79ec3f110ee2f9d38fa7d5194937e/streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java#L174-L185

AbstractProcessorContext is returning `0` when record context is null—as default value I assume. Wondering if we should align the defaults to zero, or `-1`. 

> If you wanted to do a "proper" fix, you could insert a method like invokeProcessor(Record), which would first set the timestamp on the context and then invoke the Processor.

I'm failing to see how this `invokeProcessor` should be defined. If as part of `KStreamSessionWindowAggregateProcessorTest`, I understand `setup` is already handling this initialiazation before processing—though I might be misunderstanding the proposed fixed.

Reply:
Hmm, that comment indicates that it's doing this on purpose for backwards compatibility, but now I'm wondering if I messed something up. I know that I used to have two separate implementations for the old and new context, and that later I converged them. I might have lost this backwards-compatibility feature along the way, and maybe we're just always returning `0` now. I hope that I would have anticipated that and put in a regression test, though...

Regardless, the desired behavior (now) is actually that we use -1 as a sentinel "invalid" timestamp.

Let's not worry about that "proper" fix for now. I was just suggesting that you add a helper method to this test that all the test methods would invoke instead of directly calling `process` that helper would first `setTimestamp` and then `process`. But it's really not that important.